src/queue.rs

```diff
@@ -73,29 +64,25 @@ impl MessageQueue {
                 return None;
             }
 
-            if self.nonce_set.contains(&msg.nonce) {
+            if !self.queue.insert(msg) {
                 // this shouldn't happen normally, only if the other node
                 // is not following the protocol
                 warn!("received a message with a duplicate nonce");
                 return None;
             }
 
-            self.nonce_set.insert(msg.nonce);
-            self.queue.push_back(msg);
-            self.queue.make_contiguous().sort();
             None
         }
     }
 
     pub(crate) fn pop(&mut self) -> Option<TransportMessage> {
-        let Some(head) = self.queue.front() else {
+        let Some(head) = self.queue.first() else {
             return None;
         };
 
         if head.nonce == self.next_expected_nonce {
             self.next_expected_nonce += 1;
```

Comment:
I think this needs to use `wrapping_add`
```
self.next_expected_nonce = self.next_expected_nonce.wrapping_add(1);
```

Reply:
done